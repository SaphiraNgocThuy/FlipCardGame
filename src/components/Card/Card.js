import React, { useCallback, useMemo, memo, useEffect } from 'react';
import { Animated, TouchableWithoutFeedback } from 'react-native';
import PropTypes from 'prop-types';
import {
  CardContainer,
  BackCard,
  FrontCard,
  NumberText,
  StyledImage,
} from './CardStyles';
import images from '../../assets';

const perspective = 500;

const Card = ({ title, id, isOpened, onCardPress, onFlipOpened }) => {
  const animatedValue = useMemo(() => new Animated.Value(0), []);

  useEffect(() => {
    if (isOpened) {
      Animated.spring(animatedValue, {
        toValue: 180,
        friction: 4,
        tension: 10,
      }).start();
      onFlipOpened();
    } else {
      Animated.spring(animatedValue, {
        toValue: 0,
        friction: 4,
        tension: 30,
      }).start();
    }
  }, [isOpened]);

  const frontInterpolate = useMemo(
    () =>
      animatedValue.interpolate({
        inputRange: [0, 180],
        outputRange: ['180deg', '360deg'],
      }),
    []
  );
  const backInterpolate = useMemo(
    () =>
      animatedValue.interpolate({
        inputRange: [0, 180],
        outputRange: ['0deg', '180deg'],
      }),
    []
  );

  const onPress = useCallback(() => {
    if (!isOpened) {
      onCardPress(title, id);
    }
  }, [isOpened, title, id]);

  return (
    <TouchableWithoutFeedback {...{ onPress }} testID="touchable">
      <CardContainer>
        <FrontCard
          style={{
            transform: [
              { perspective },
              {
                rotateY: frontInterpolate,
              },
            ],
          }}
        >
          <NumberText>{title}</NumberText>
        </FrontCard>
        <BackCard
          style={{
            transform: [
              { perspective },
              {
                rotateY: backInterpolate,
              },
            ],
          }}
        >
          <StyledImage source={images.fruit.strawberry} />
        </BackCard>
      </CardContainer>
    </TouchableWithoutFeedback>
  );
};

Card.propTypes = {
  id: PropTypes.number.isRequired,
  title: PropTypes.number.isRequired,
  isOpened: PropTypes.bool.isRequired,
  onCardPress: PropTypes.func.isRequired,
  onFlipOpened: PropTypes.func.isRequired,
};

export default memo(Card);
