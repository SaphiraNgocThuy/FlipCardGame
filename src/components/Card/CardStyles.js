import styled from 'styled-components/native';
import { theme } from '../../utils/theme';
import { padding, fontSize, borderRadius } from '../../utils/commonStyles';
import { Animated } from 'react-native';

export const CardContainer = styled.View`
  flex-grow: 1;
  flex-basis: 25%;
  margin: ${padding.sm}px;
`;

const MainCard = styled(Animated.View)`
  position: absolute;
  backface-visibility: hidden;
  width: 100%;
  height: 100%;
  border-radius: ${borderRadius}px;
  justify-content: center;
  align-items: center;
`;

export const FrontCard = styled(MainCard)`
  background-color: ${theme.primary};
`;

export const BackCard = styled(MainCard)`
  background-color: ${theme.darkGrey};
`;

export const NumberText = styled.Text`
  font-size: ${fontSize.xl}px;
  color: ${theme.bright};
`;

export const StyledImage = styled.Image.attrs({
  resizeMode: 'contain',
})`
  width: 80%;
  height: 80%;
`;
