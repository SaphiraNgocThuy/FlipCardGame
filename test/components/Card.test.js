import React from 'react';
import renderer from 'react-test-renderer';
import Card from '../../src/components/Card/Card';
import { render, fireEvent } from 'react-native-testing-library';
import images from '../../src/assets';

it('renders when not open', () => {
  const tree = renderer
    .create(
      <Card
        title={1}
        id={1}
        isOpened={false}
        onCardPress={() => {}}
        onFlipOpened={() => {}}
        background={images.fruit.strawberry}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders if opened', () => {
  const tree = renderer
    .create(
      <Card
        title={1}
        id={1}
        isOpened={true}
        onCardPress={() => {}}
        onFlipOpened={() => {}}
        background={images.fruit.strawberry}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('onPress when card is not open', () => {
  const onCardPress = jest.fn();
  const rendered = render(
    <Card
      title={1}
      id={1}
      isOpened={false}
      {...{ onCardPress }}
      onFlipOpened={() => {}}
      background={images.fruit.strawberry}
    />
  );
  const button = rendered.getByTestId('touchable');
  fireEvent(button, 'press');
  expect(onCardPress).toHaveBeenCalledWith(1, 1);
});

it('onPress when card is open', () => {
  const onCardPress = jest.fn();
  const rendered = render(
    <Card
      title={1}
      id={1}
      isOpened
      {...{ onCardPress }}
      onFlipOpened={() => {}}
      background={images.fruit.strawberry}
    />
  );
  const button = rendered.getByTestId('touchable');
  fireEvent(button, 'press');
  expect(onCardPress).not.toHaveBeenCalled();
});
