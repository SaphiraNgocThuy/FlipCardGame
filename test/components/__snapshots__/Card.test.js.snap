// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders if opened 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "flexBasis": "25%",
        "flexGrow": 1,
        "marginBottom": 16,
        "marginLeft": 16,
        "marginRight": 16,
        "marginTop": 16,
      },
    ]
  }
  testID="touchable"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#33c4b3",
        "borderRadius": 60,
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "transform": Array [
          Object {
            "perspective": 500,
          },
          Object {
            "rotateY": "180deg",
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 60,
          },
        ]
      }
    >
      1
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#424c58",
        "borderRadius": 60,
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "transform": Array [
          Object {
            "perspective": 500,
          },
          Object {
            "rotateY": "0deg",
          },
        ],
        "width": "100%",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={1}
      style={
        Array [
          Object {
            "height": "80%",
            "width": "80%",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`renders when not open 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "flexBasis": "25%",
        "flexGrow": 1,
        "marginBottom": 16,
        "marginLeft": 16,
        "marginRight": 16,
        "marginTop": 16,
      },
    ]
  }
  testID="touchable"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#33c4b3",
        "borderRadius": 60,
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "transform": Array [
          Object {
            "perspective": 500,
          },
          Object {
            "rotateY": "180deg",
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 60,
          },
        ]
      }
    >
      1
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#424c58",
        "borderRadius": 60,
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "transform": Array [
          Object {
            "perspective": 500,
          },
          Object {
            "rotateY": "0deg",
          },
        ],
        "width": "100%",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={1}
      style={
        Array [
          Object {
            "height": "80%",
            "width": "80%",
          },
        ]
      }
    />
  </View>
</View>
`;
